### PR TITLE
Only replace FECollection in *DoFHandler if necessary

### LINKED ISSUE #5715: Bug using MappingFEField and refinements
Working with @luca-heltai we discover that the following code fails. We don't know yet who or what introduced this issue. The code fails because the subscriptor of gradient_dh, which is needed from MappingFEField, is not zero when the second distribute_dofs comes in place. With version 8.5 the code is functioning just fine.

```
#include <deal.II/base/utilities.h>
#include <deal.II/base/quadrature_lib.h>
#include <deal.II/base/geometry_info.h>
#include<deal.II/dofs/dof_handler.h>
#include<deal.II/dofs/dof_accessor.h>
#include<deal.II/dofs/dof_tools.h>
#include<deal.II/fe/fe_q.h>
#include<deal.II/fe/fe_values.h>
#include<deal.II/fe/fe_system.h>
#include<deal.II/fe/mapping_q1_eulerian.h>
#include<deal.II/fe/mapping_q1.h>
#include <deal.II/fe/mapping_fe_field.h>
#include <deal.II/grid/grid_generator.h>
#include<deal.II/numerics/vector_tools.h>

using namespace dealii;
int main()
{


  Triangulation<2> tria;
  GridGenerator::hyper_cube (tria);
  tria.refine_global (1);
  FE_Q<2> fe(1);
  FESystem<2> gradient_fe(FE_Q<2> (1),1,FE_Q<2> (1),1);
  DoFHandler<2,2> dh(tria);
  DoFHandler<2,2> gradient_dh(tria);
  dh.distribute_dofs(fe);
  gradient_dh.distribute_dofs(gradient_fe);
  std::cout<<tria.n_active_cells()<<" "<<dh.n_dofs()<<std::endl;
  Vector<double> map_vector;
  map_vector.reinit(gradient_dh.n_dofs());
  VectorTools::get_position_vector(gradient_dh, map_vector);
  MappingFEField<2, 2> mapping(gradient_dh, map_vector);
  tria.refine_global (1);
  dh.distribute_dofs(fe);
  gradient_dh.distribute_dofs(gradient_fe);
  std::cout<<tria.n_active_cells()<<" "<<dh.n_dofs()<<std::endl;
}
```

### LINKED ISSUE #6325: Bug in DoFHandler and hp::DoFHandler
The following minimal example triggers an exception

~~~
Triangulation<dim,spacedim> tria;
  GridGenerator::hyper_cube (tria);
  FE_Q<dim,spacedim> fe(1);

  DoFHandler<dim,spacedim> dh(tria);
  dh.distribute_dofs(fe);

  // This stores a pointer to the fe in dh.
  FEValues<dim,spacedim> fe_v(dh.get_fe(), QGauss<dim>(1), update_values);

  tria.refine_global (1);
  dh.distribute_dofs(fe);
~~~
The second `distribute_dofs` will fail, saying that someone is still using the fe.

Notice that if we use `fe` instead of `dh.get_fe()`, the test passes.

Should we **clone** the fe also inside `FEValues`?

### PR DESCRIPTION
**Pull request by masterleinad**
Supersedes #6326 and #6336 and fixes #6325, fixes #5715.

### CODE REVIEW

File: source/dofs/dof_handler.cc [hunk 1]
```diff
@@ -1000,7 +1000,7 @@ void DoFHandler<dim,spacedim>::distribute_dofs (const FiniteElement<dim,spacedim
 
   // Only recreate the FECollection if we don't already store
   // the exact same FiniteElement object.
-  if  (fe_collection.size() == 0 || &(fe_collection[0]) != &ff)
```
Comment (davydden) [inline source/dofs/dof_handler.cc]: that might not work for FE_Enriched as in the most general case `const std::vector< std::vector< std::function<...>>>` is not encoded in its [name](https://www.dealii.org/developer/doxygen/deal.II/fe__enriched_8cc_source.html#l00429), but i guess it's the problem of FE_Enriched.

File: source/hp/dof_handler.cc [hunk 2]
```diff
@@ -1284,7 +1284,18 @@ namespace hp
     Assert (ff.size() > 0,
             ExcMessage("The hp::FECollection given is empty!"));
 
-    if (&fe_collection != &ff)
+    // don't create a new object if the one we have is already appropriate
+    bool is_same = (fe_collection.size() == ff.size());
```
Comment (davydden) [inline source/hp/dof_handler.cc]: one could as well add `operator==` for `FECollection`, that could be useful elsewhere.

**Comment by bangerth** [comment]
There are other elements that may have internal state that is not reflected in the name. For example, `FE_Nothing` has a flag that is given to the constructor, but I think is not reflected in the name. I'm also not sure if `FE_Q` initialized with a quadrature formula tells all of this in its name.

The name does not seem like a reliable criterion to compare elements. What if we called `operator==` instead?

**Comment by luca-heltai** [comment]
Ok to merge once the tester is happy.

File: include/deal.II/hp/fe_collection.h [hunk 3]
```diff
@@ -495,6 +501,27 @@ namespace hp
...
+  FECollection<dim,spacedim>::operator== (const FECollection<dim,spacedim> &fe_collection) const
+  {
+    bool is_same = (size() == fe_collection.size());
+    if (is_same)
+      {
+        for (unsigned int i=0; i<size(); ++i)
+          if (!(*finite_elements[i] == fe_collection[i]))
+            {
+              is_same = false;
+              break;
```
Comment (bangerth) [inline include/deal.II/hp/fe_collection.h]: just say `return false` here.

File: include/deal.II/hp/fe_collection.h [hunk 4]
```diff
@@ -495,6 +501,27 @@ namespace hp
...
+  template <int dim, int spacedim>
+  inline
+  bool
+  FECollection<dim,spacedim>::operator== (const FECollection<dim,spacedim> &fe_collection) const
+  {
+    bool is_same = (size() == fe_collection.size());
+    if (is_same)
+      {
+        for (unsigned int i=0; i<size(); ++i)
+          if (!(*finite_elements[i] == fe_collection[i]))
```
Comment (bangerth) [inline include/deal.II/hp/fe_collection.h]: Hm, this currently only calls a non-`virtual` `operator==` in the `FiniteElement` case class. I wonder if that is going to call two elements the same that are really not. For example, the `FE_Nothing(1,true)` and `FE_Nothing(1,false)` are almost certainly going to be called the same, though they are not :-(

File: include/deal.II/hp/fe_collection.h [hunk 5]
```diff
@@ -100,6 +100,12 @@ namespace hp
     FECollection<dim, spacedim> &
     operator= (FECollection<dim,spacedim> &&) = default; // NOLINT
 
+    /**
+     * Comparison operator. All stored Finite Element objects are compared in order.
+     */
+    bool
+    operator== (const FECollection<dim,spacedim> &fe_collection) const;
```
Comment (davydden) [inline include/deal.II/hp/fe_collection.h]: while there, could you please add `operator !=`? That would make code below read better.

File: include/deal.II/hp/fe_collection.h (hunk 4 above)
Comment (davydden) [reply include/deal.II/hp/fe_collection.h]: >  the FE_Nothing(1,true) and FE_Nothing(1,false) are almost certainly going to be called the same, though they are not :-(

yeah... If we merge this PR, we should at least create an issue to investigate this issue later...

**Comment by bangerth** [comment]
I think we got ourselves into a rabbit hole that we can't find our way out of any more in the time we have.

What would people think of reverting the original patch? I'm not opposed to the general idea, but I'm afraid that we did not think this through and will find cases that don't work after any imminent release...

**Comment by davydden** [comment]
> the original patch

what do you mean by original patch?

**Comment by masterleinad** [comment]
> What would people think of reverting the original patch? I'm not opposed to the general idea, but I'm afraid that we did not think this through and will find cases that don't work after any imminent release...

In almost all cases this patch is going to work, but of course this is not enough. I can imagine that `FE_Nothing` is the only class for which this might lead to wrong results if one would call `distribute_dofs` with `FE_Nothing(1,true)` and after that with `FE_Nothing(1,true)` as argument.

I can try to get an alternative using `SmartPointer`s again.

**Comment by luca-heltai** [comment]
What was the original reason to **clone** the fe? If the reason is a valid one, than maybe we should just merge this, and fix the "==" problems later on.

**Comment by masterleinad** [comment]
Currently, we are also checking equality of `FIniteElement::get_name()` in the comparison operator. I went through our `FiniteElement` classes and all of them (after fixing `FE_Nothing`) should be uniquely determined by their name apart from `FE_DGQArbitraryNodes`, `FE_Q`, `FE_Q_Bubbles` and `FE_Q_DG0`. All of these can be initialized either by a degree or by support points given by `Quadrature` parameter. We currently identify equidistant support points and `QGaussLobatto` correctly for `FE_Q`, `FE_Q_Bubbles` and `FE_Q_DG0`. For `FE_DGQArbitraryNodes` we additionally recognize `QGauss` and `QGaussLog`.
I guess this is more or less (maybe extending the number of sets of support points recognized) how far we can get with this approach.

**Comment by bangerth** [comment]
You'll have to update the output of `hp/hp_constraints_neither_dominate_01`.

I think this could work. I'm still convinced that we can find cases where `operator==` incorrectly returns `true`, but we can address them in follow-ups. Does this pass all of the tests people have suggested on how to break the previous version?

**Comment by davydden** [comment]
@masterleinad 

> all of them (after fixing FE_Nothing) should be uniquely determined by their name apart from FE_DGQArbitraryNodes, FE_Q, FE_Q_Bubbles and FE_Q_DG0. 

you need to add `FE_Enriched` to that list.

@bangerth 

> I'm still convinced that we can find cases where operator== incorrectly returns true

yeap, FE_Enriched with two different enrichment functions. More generally the enrichments are given by
```
const std::vector< std::vector< std::function< const Function< spacedim > *(const typename Triangulation< dim, spacedim >::cell_iterator &)>>> &
```

But I also agree that it should be possible to fix it.

**Comment by masterleinad** [comment]
> Does this pass all of the tests people have suggested on how to break the previous version?

I think so. I also included the test in #6336.

Additionally, I restructured the commits a bit.

File: doc/news/changes/minor/20180425DanielArndt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+New: FECollection gained a equality comparison operator. Both FiniteElement
```
Comment (kronbichler) [inline doc/news/changes/minor/20180425DanielArndt]: a -> an

**Pull request merged**